File: editor/src/parser.ts
/**
 * VNA Parser for TypeScript/JavaScript
 * This parser is designed to work in the editor environment
 */

export interface Metadata {
  title: string;
  raga: string;
  tala: string;
  tempo?: number;
  composer?: string;
  language?: string;
  type?: string;
  key?: string;
  gati?: number;
  default_octave?: string;
  arohanam?: string;
  avarohanam?: string;
}

export interface Phrase {
  swaras: string[];
  sahitya: string[];
  phrase_analysis?: string;
  line_number: number;
  beat_positions: number[];
  gati?: number;
}

export interface Section {
  name: string;
  phrases: Phrase[];
  line_number: number;
  gati?: number;
}

export interface VnaDocument {
  metadata: Metadata;
  sections: Section[];
}

export interface ValidationIssue {
  severity: 'error' | 'warning' | 'info';
  message: string;
  line: number;
  column?: number;
}

export class VnaParser {
  private lines: string[];
  private currentLine: number;

  constructor() {
    this.lines = [];
    this.currentLine = 0;
  }

  parse(content: string): VnaDocument {
    this.lines = content.split('\n');
    this.currentLine = 0;

    const metadata = this.parseMetadata();
    const sections = this.parseSections();

    return { metadata, sections };
  }

  private parseMetadata(): Metadata {
    if (!this.currentLineStartsWith('---')) {
      throw new Error('Missing YAML frontmatter');
    }

    this.advanceLine(); // Skip opening ---
    const yamlLines: string[] = [];

    while (this.currentLine < this.lines.length) {
      const line = this.lines[this.currentLine];
      if (line.trim() === '---') {
        this.advanceLine();
        break;
      }
      yamlLines.push(line);
      this.advanceLine();
    }

    // Simple YAML parsing (for basic key-value pairs)
    const metadata: any = {};
    for (const line of yamlLines) {
      const match = line.match(/^(\w+):\s*"?([^"]*)"?$/);
      if (match) {
        const [, key, value] = match;
        if (key === 'tempo' || key === 'gati') {
          metadata[key] = parseInt(value, 10);
        } else {
          metadata[key] = value;
        }
      }
    }

    // Validate required fields
    if (!metadata.title || !metadata.raga || !metadata.tala) {
      throw new Error('Missing required metadata fields');
    }

    return metadata as Metadata;
  }

  private parseSections(): Section[] {
    const sections: Section[] = [];

    while (this.currentLine < this.lines.length) {
      const line = this.currentLineTrimmed();

      if (line.startsWith('[') && line.endsWith(']')) {
        const section = this.parseSection();
        sections.push(section);
      } else if (line.trim() !== '' && !line.startsWith('#')) {
        throw new Error(`Unexpected content at line ${this.currentLine + 1}`);
      } else {
        this.advanceLine();
      }
    }

    return sections;
  }

  private parseSection(): Section {
    const line = this.currentLineTrimmed();
    const sectionLine = this.currentLine;

    const name = line.slice(1, -1);
    this.advanceLine();

    const phrases: Phrase[] = [];
    let sectionGati: number | undefined;

    while (this.currentLine < this.lines.length) {
      const line = this.currentLineTrimmed();

      if (line === '') {
        this.advanceLine();
        continue;
      }

      if (line.startsWith('#')) {
        this.advanceLine();
        continue;
      }

      if (line.startsWith('@gati:')) {
        sectionGati = parseInt(line.substring(6).trim(), 10);
        this.advanceLine();
        continue;
      }

      if (line.startsWith('[') && line.endsWith(']')) {
        break;
      }

      if (line.includes('|')) {
        const phrase = this.parsePhrase();
        phrases.push(phrase);
      } else {
        throw new Error(`Unexpected content in section at line ${this.currentLine + 1}`);
      }
    }

    return {
      name,
      phrases,
      line_number: sectionLine + 1,
      gati: sectionGati,
    };
  }

  private parsePhrase(): Phrase {
    const phraseStartLine = this.currentLine;
    let lineGati: number | undefined;

    // Check for line-level gati
    if (this.currentLineTrimmed().startsWith('@gati:')) {
      lineGati = parseInt(this.currentLineTrimmed().substring(6).trim(), 10);
      this.advanceLine();
    }

    // Parse swara line
    const swaraLine = this.currentLineTrimmed();
    const { tokens: swaras, beatPositions: swaraBeatPositions } = this.parseNotationLine(swaraLine);
    this.advanceLine();

    // Parse sahitya line
    const sahityaLine = this.currentLineTrimmed();
    const { tokens: sahitya, beatPositions: sahityaBeatPositions } = this.parseNotationLine(sahityaLine);
    this.advanceLine();

    // Verify beat alignment
    if (JSON.stringify(swaraBeatPositions) !== JSON.stringify(sahityaBeatPositions)) {
      throw new Error(`Beat markers misaligned at line ${phraseStartLine + 1}`);
    }

    // Check for phrase analysis
    let phraseAnalysis: string | undefined;
    if (this.currentLineTrimmed().startsWith('phrases = ')) {
      phraseAnalysis = this.currentLineTrimmed().substring(10);
      this.advanceLine();
    }

    return {
      swaras,
      sahitya,
      phrase_analysis: phraseAnalysis,
      line_number: phraseStartLine + 1,
      beat_positions: swaraBeatPositions,
      gati: lineGati,
    };
  }

  private parseNotationLine(line: string): { tokens: string[]; beatPositions: number[] } {
    // Remove || at end
    const cleanLine = line.endsWith('||') ? line.slice(0, -2).trim() : line.trim();

    const tokens: string[] = [];
    const beatPositions: number[] = [];
    let currentPos = 0;

    // Split by | to get beats
    const beats = cleanLine.split('|');

    beats.forEach((beat, i) => {
      const beatTokens = beat.trim().split(/\s+/).filter(t => t !== '');
      beatTokens.forEach(token => {
        tokens.push(token);
        currentPos++;
      });

      // Record beat position after this beat (except for last beat)
      if (i < beats.length - 1 && currentPos > 0) {
        beatPositions.push(currentPos);
      }
    });

    return { tokens, beatPositions };
  }

  private currentLineTrimmed(): string {
    return this.currentLine < this.lines.length ? this.lines[this.currentLine].trim() : '';
  }

  private currentLineStartsWith(prefix: string): boolean {
    return this.currentLineTrimmed().startsWith(prefix);
  }

  private advanceLine(): void {
    this.currentLine++;
  }

  // Validation
  validate(content: string): ValidationIssue[] {
    const issues: ValidationIssue[] = [];

    try {
      const document = this.parse(content);
      
      // Validate metadata
      if (document.metadata.tempo && (document.metadata.tempo < 20 || document.metadata.tempo > 300)) {
        issues.push({
          severity: 'warning',
          message: `Unusual tempo: ${document.metadata.tempo} BPM (typical range: 20-300)`,
          line: 1,
        });
      }

      if (document.metadata.gati && ![3, 4, 5, 7, 9].includes(document.metadata.gati)) {
        issues.push({
          severity: 'warning',
          message: `Unusual gati value: ${document.metadata.gati} (typical values: 3, 4, 5, 7, 9)`,
          line: 1,
        });
      }

      // Validate phrases
      for (const section of document.sections) {
        for (const phrase of section.phrases) {
          // Check token count
          if (phrase.swaras.length !== phrase.sahitya.length) {
            issues.push({
              severity: 'error',
              message: `Token count mismatch: ${phrase.swaras.length} swaras vs ${phrase.sahitya.length} sahitya`,
              line: phrase.line_number,
            });
            continue;
          }

          // Check token length matching
          for (let i = 0; i < phrase.swaras.length; i++) {
            const swaraLen = phrase.swaras[i].length;
            const sahityaLen = phrase.sahitya[i].length;
            if (swaraLen !== sahityaLen) {
              issues.push({
                severity: 'error',
                message: `Token length mismatch at position ${i + 1}: "${phrase.swaras[i]}" (${swaraLen}) vs "${phrase.sahitya[i]}" (${sahityaLen})`,
                line: phrase.line_number + 1,
              });
            }
          }
        }
      }
    } catch (e) {
      issues.push({
        severity: 'error',
        message: e instanceof Error ? e.message : 'Unknown parsing error',
        line: this.currentLine + 1,
      });
    }

    return issues;
  }

  // Utility functions for editor features
  getSwaraTokens(): string[] {
    return ['S', 'R', 'G', 'M', 'P', 'D', 'N',
            'R1', 'R2', 'R3', 'G1', 'G2', 'G3',
            'M1', 'M2', 'D1', 'D2', 'D3', 'N1', 'N2', 'N3'];
  }

  getSectionNames(): string[] {
    return ['pallavi', 'anupallavi', 'muktasvara', 'charanam',
            'cittasvaras', 'geetam', 'ragamalika', 'madhyamakala', 'drut'];
  }

  getMetadataFields(): Array<{ name: string; required: boolean }> {
    return [
      { name: 'title', required: true },
      { name: 'raga', required: true },
      { name: 'tala', required: true },
      { name: 'tempo', required: false },
      { name: 'composer', required: false },
      { name: 'language', required: false },
      { name: 'type', required: false },
      { name: 'key', required: false },
      { name: 'gati', required: false },
      { name: 'default_octave', required: false },
      { name: 'arohanam', required: false },
      { name: 'avarohanam', required: false },
    ];
  }
}

File: vna-lsp/src/parser.rs
use crate::types::*;
use anyhow::{anyhow, Result};
use serde_yaml;

pub fn parse(content: &str) -> Result<VnaDocument> {
    let mut parser = VnaParser::new(content);
    parser.parse()
}

struct VnaParser {
    content: String,
    lines: Vec<String>,
    current_line: usize,
}

impl VnaParser {
    fn new(content: &str) -> Self {
        let lines: Vec<String> = content.lines().map(|l| l.to_string()).collect();
        Self {
            content: content.to_string(),
            lines,
            current_line: 0,
        }
    }

    fn parse(&mut self) -> Result<VnaDocument> {
        let metadata = self.parse_metadata()?;
        let (sections, comments) = self.parse_body()?;

        Ok(VnaDocument {
            metadata,
            sections,
            comments,
        })
    }

    fn parse_metadata(&mut self) -> Result<Metadata> {
        // Look for YAML frontmatter
        if !self.current_line_starts_with("---") {
            return Err(anyhow!("Missing YAML frontmatter at start of file"));
        }

        self.advance_line(); // Skip opening ---
        let mut yaml_lines = Vec::new();
        
        while self.current_line < self.lines.len() {
            let line = &self.lines[self.current_line];
            if line.trim() == "---" {
                self.advance_line(); // Skip closing ---
                break;
            }
            yaml_lines.push(line.clone());
            self.advance_line();
        }

        if yaml_lines.is_empty() {
            return Err(anyhow!("Empty YAML frontmatter"));
        }

        let yaml_content = yaml_lines.join("\n");
        let metadata: Metadata = serde_yaml::from_str(&yaml_content)
            .map_err(|e| anyhow!("Invalid YAML metadata: {}", e))?;

        // Validate required fields
        if metadata.title.is_empty() {
            return Err(anyhow!("Missing required field: title"));
        }
        if metadata.raga.is_empty() {
            return Err(anyhow!("Missing required field: raga"));
        }
        if metadata.tala.is_empty() {
            return Err(anyhow!("Missing required field: tala"));
        }

        Ok(metadata)
    }

    fn parse_body(&mut self) -> Result<(Vec<Section>, Vec<Comment>)> {
        let mut sections = Vec::new();
        let mut comments = Vec::new();

        while self.current_line < self.lines.len() {
            let line = self.current_line_trimmed();
            
            if line.is_empty() {
                self.advance_line();
                continue;
            }

            if line.starts_with('#') {
                comments.push(Comment {
                    text: line[1..].trim().to_string(),
                    line_number: self.current_line + 1,
                    comment_type: CommentType::Line,
                });
                self.advance_line();
                continue;
            }

            if line.starts_with('[') && line.ends_with(']') {
                let section = self.parse_section()?;
                sections.push(section);
                continue;
            }

            return Err(anyhow!(
                "Unexpected content at line {}: {}",
                self.current_line + 1,
                line
            ));
        }

        Ok((sections, comments))
    }

    fn parse_section(&mut self) -> Result<Section> {
        let line = self.current_line_trimmed();
        let section_line = self.current_line;
        
        if !line.starts_with('[') || !line.ends_with(']') {
            return Err(anyhow!("Invalid section header at line {}", section_line + 1));
        }

        let name = line[1..line.len()-1].to_string();
        self.advance_line();

        let mut phrases = Vec::new();
        let mut pending_comments = Vec::new();
        let mut section_comments = Vec::new();
        let mut section_gati = None;
        let mut section_tala = None;

        while self.current_line < self.lines.len() {
            let line = self.current_line_trimmed();
            
            // Empty line - continue
            if line.is_empty() {
                self.advance_line();
                continue;
            }

            // Comment - collect
            if line.starts_with('#') {
                let comment = Comment {
                    text: line[1..].trim().to_string(),
                    line_number: self.current_line + 1,
                    comment_type: CommentType::Line,
                };
                pending_comments.push(comment);
                self.advance_line();
                continue;
            }

            // New section - break
            if line.starts_with('[') && line.ends_with(']') {
                // Any pending comments belong to the section
                section_comments.extend(pending_comments.drain(..));
                break;
            }

            // Gati annotation
            if line.starts_with("@gati:") {
                // Section-level gati override
                let gati_str = line[6..].trim();
                if let Ok(gati) = gati_str.parse::<u8>() {
                    section_gati = Some(gati);
                } else {
                    return Err(anyhow!("Invalid gati value at line {}: {}", self.current_line + 1, gati_str));
                }
                self.advance_line();
                continue;
            }

            // Tala annotation
            if line.starts_with("@tala:") {
                // Section-level tala override
                let tala_str = line[6..].trim().trim_matches('"');
                section_tala = Some(tala_str.to_string());
                self.advance_line();
                continue;
            }

            // Notation line - parse phrase
            if line.contains('|') {
                let mut phrase = self.parse_phrase()?;
                phrase.preceding_comments = pending_comments.drain(..).collect();
                phrases.push(phrase);
                continue;
            }

            return Err(anyhow!(
                "Unexpected content in section '{}' at line {}: {}",
                name,
                self.current_line + 1,
                line
            ));
        }

        // Any remaining comments belong to the section
        section_comments.extend(pending_comments);

        Ok(Section {
            name,
            phrases,
            line_number: section_line + 1,
            comments: section_comments,
            gati: section_gati,
            tala: section_tala,
        })
    }

    fn parse_phrase(&mut self) -> Result<Phrase> {
        let phrase_start_line = self.current_line;
        let mut line_gati = None;
        let mut line_tala = None;

        // Check for line-level annotations before the swara line
        while self.current_line < self.lines.len() {
            let line = self.current_line_trimmed();
            
            if line.starts_with("@gati:") {
                let gati_str = line[6..].trim();
                if let Ok(gati) = gati_str.parse::<u8>() {
                    line_gati = Some(gati);
                } else {
                    return Err(anyhow!("Invalid gati value at line {}: {}", self.current_line + 1, gati_str));
                }
                self.advance_line();
            } else if line.starts_with("@tala:") {
                let tala_str = line[6..].trim().trim_matches('"');
                line_tala = Some(tala_str.to_string());
                self.advance_line();
            } else {
                break; // Not an annotation, must be the swara line
            }
        }

        // Expect at least 2 lines: swara, sahitya
        if self.current_line + 1 >= self.lines.len() {
            return Err(anyhow!(
                "Incomplete phrase at line {} - need at least 2 lines (swara, sahitya)",
                phrase_start_line + 1
            ));
        }

        // Parse swara line
        let swara_line = self.current_line_trimmed();
        if !swara_line.contains('|') {
            return Err(anyhow!(
                "Invalid swara line at {}: missing beat markers",
                self.current_line + 1
            ));
        }
        let (swaras, swara_beats) = self.parse_notation_line_with_beats(&swara_line)?;
        self.advance_line();

        // Parse sahitya line
        let sahitya_line = self.current_line_trimmed();
        if !sahitya_line.contains('|') {
            return Err(anyhow!(
                "Invalid sahitya line at {}: missing beat markers",
                self.current_line + 1
            ));
        }
        let (sahitya, sahitya_beats) = self.parse_notation_line_with_beats(&sahitya_line)?;
        self.advance_line();

        // Check for optional phrase analysis line
        let mut phrase_analysis = None;
        if self.current_line < self.lines.len() {
            let next_line = self.current_line_trimmed();
            if next_line.starts_with("phrases = ") {
                phrase_analysis = Some(next_line[10..].to_string());
                self.advance_line();
            }
        }

        // Verify beat alignment
        if swara_beats != sahitya_beats {
            return Err(anyhow!(
                "Beat markers misaligned between swara and sahitya lines at line {}",
                phrase_start_line + 1
            ));
        }

        Ok(Phrase {
            swaras,
            sahitya,
            phrase_analysis,
            line_number: phrase_start_line + 1,
            preceding_comments: Vec::new(), // Will be filled by parse_section
            gati: line_gati,
            tala: line_tala,
            beat_positions: swara_beats,
        })
    }

    fn parse_notation_line(&self, line: &str) -> Result<Vec<String>> {
        let (elements, _) = self.parse_notation_line_with_beats(line)?;
        Ok(elements)
    }

    fn parse_notation_line_with_beats(&self, line: &str) -> Result<(Vec<String>, Vec<usize>)> {
        // Remove || at end
        let clean_line = if line.ends_with("||") {
            &line[..line.len() - 2]
        } else {
            line
        }.trim();
        
        let mut elements = Vec::new();
        let mut beat_positions = Vec::new();
        let mut current_pos = 0;
        
        // Split by | to get beats
        let beats: Vec<&str> = clean_line.split('|').collect();
        
        for (i, beat) in beats.iter().enumerate() {
            let beat_elements: Vec<&str> = beat.trim().split_whitespace().collect();
            for element in beat_elements {
                if !element.is_empty() {
                    elements.push(element.to_string());
                    current_pos += 1;
                }
            }
            
            // Record beat position after this beat (except for last beat)
            if i < beats.len() - 1 && current_pos > 0 {
                beat_positions.push(current_pos);
            }
        }

        Ok((elements, beat_positions))
    }

    fn current_line_trimmed(&self) -> String {
        if self.current_line < self.lines.len() {
            self.lines[self.current_line].trim().to_string()
        } else {
            String::new()
        }
    }

    fn current_line_starts_with(&self, prefix: &str) -> bool {
        if self.current_line < self.lines.len() {
            self.lines[self.current_line].trim().starts_with(prefix)
        } else {
            false
        }
    }

    fn advance_line(&mut self) {
        self.current_line += 1;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_complete_file() {
        let content = r#"---
title: "Test Varnam"
raga: "mohanam"
tala: "+234+0+0"
tempo: 60
---

[pallavi]
G , G , | R , , , | S S R R ||
nin - nu - | ko - - - | ri - - - ||

[anupallavi]
P D S' D | P G R S ||
pa da sa da | pa ga ra sa ||
"#;

        let result = parse(content);
        assert!(result.is_ok());
        
        let doc = result.unwrap();
        assert_eq!(doc.metadata.title, "Test Varnam");
        assert_eq!(doc.metadata.raga, "mohanam");
        assert_eq!(doc.sections.len(), 2);
        assert_eq!(doc.sections[0].name, "pallavi");
        assert_eq!(doc.sections[0].phrases.len(), 1);
    }

    #[test]
    fn test_parse_invalid_yaml() {
        let content = r#"---
title: "Test"
invalid_yaml: [unclosed
---

[pallavi]
G ||
nin ||
"#;
        
        let result = parse(content);
        assert!(result.is_err());
    }

    #[test]
    fn test_parse_missing_required_field() {
        let content = r#"---
title: "Test"
# Missing raga and tala
---

[pallavi]
G ||
nin ||
"#;
        
        let result = parse(content);
        assert!(result.is_err());
    }

    #[test]
    fn test_parse_incomplete_phrase() {
        let content = r#"---
title: "Test"
raga: "mohanam"
tala: "+234+0+0"
---

[pallavi]
G , G , | R , , , ||
# Missing sahitya line
"#;
        
        let result = parse(content);
        assert!(result.is_err());
    }

    #[test]
    fn test_parse_with_phrase_analysis() {
        let content = r#"---
title: "Test"
raga: "mohanam"
tala: "+234+0+0"
---

[pallavi]
G , G , | R , , , ||
nin - nu - | ko - - - ||
phrases = (_ *)* *   * *
"#;
        
        let result = parse(content);
        assert!(result.is_ok());
        let doc = result.unwrap();
        assert_eq!(doc.sections[0].phrases[0].phrase_analysis, Some("(_ *)* *   * *".to_string()));
    }
}